Integration tests for mac_shadow

twangboy opened the pull request:
### What does this PR do?

Adds integration tests for mac_shadow
### Tests written?

Yes

twangboy committed:
Add integration tests for mac_shadow

twangboy committed:
Fix __random_string assignment

twangboy committed:
Add destructiveTest decorators

cachedout commented on tests/integration/modules/mac_shadow.py:
```diff
@@ -0,0 +1,163 @@
+        '''
+        self.run_function('user.add', [TEST_USER])
+        ret = self.run_function('shadow.info', [TEST_USER])
+        self.assertTrue(isinstance(ret, dict))
```
Why assert this type when the assert right below it is pulling a key from a dict?

twangboy replied:
I will remove the dict check.

cachedout commented on tests/integration/modules/mac_shadow.py:
```diff
@@ -0,0 +1,163 @@
+        self.run_function('user.add', [TEST_USER])
+        text_date = self.run_function('shadow.get_account_created', [TEST_USER])
+        obj_date = datetime.datetime.strptime(text_date, '%Y-%m-%d %H:%M:%S')
+        self.assertTrue(isinstance(obj_date, datetime.date))
```
This doesn't make sense to me. It seems like what you're doing here is effectively checking to see if `strptime` can convert the timestamp, but if it can't, you'll raise a `ValueError` (per the documentation) and you'll never even _reach_ this assert. This problem exists throughout this test class.

twangboy replied:
This is getting the date/time the account was created. There's no way to know what date/time is going to be returned. All I can do is check that a date/time has been returned. How would you recommend testing this one?

cachedout replied:
If simply testing the date were sufficient, you could certainly assert that it was between the time that the account was created and the time of the assertion. Though, reading the code in question it's not clear to me what will happen with `get_account_created` if the user does not exist. It sort of looks like it might blow up, though I don't have an OS X box to test it on.

twangboy replied:
It returns `Value not set`. Is there a way to put the datetime thing in a try/except block and fail the test if there's an exception?

cachedout replied:
i think it's sort of weird that these functions apparently return strings both on success and on failure. Why not return a boolean in case of a failure?

cachedout commented on tests/integration/modules/mac_shadow.py:
```diff
@@ -0,0 +1,163 @@
+        Test shadow.del_password
+        '''
+        self.run_function('user.add', [TEST_USER])
+        self.assertTrue(self.run_function('shadow.del_password', [TEST_USER]))
```
This doesn't seem sufficient to me. When we're writing integration tests for execution modules, we need to assert that the system is in the state that it should be. Simply asserting that a function returned True really doesn't tell us very much.

twangboy replied:
The execution module returns true if '*' is found as the password after the password has been deleted. The check is being done there.

twangboy committed:
Total rewrite of mac_shadow.py

Updated mac_shadow integration tests

twangboy committed:
Replaced cmd.run with mac_utils execute command

twangboy committed:
Fix some pylint errors

twangboy committed:
Add test for _convert_to_datetime

twangboy committed:
Remove test for _convert_to_datetime

twangboy committed:
Fix typo

twangboy committed:
Remove redundant try/except blocks

cachedout commented:
Go Go Jenkins!

rallytime commented:
Go Go Jenkins!

twangboy commented:
Rebased

cachedout commented on salt/modules/mac_shadow.py:
```diff
@@ -97,19 +45,17 @@ def _get_account_policy(name):
+                policy_dict[key] = value
+        return policy_dict
+    except IndexError:
         return 'Value not found'
```
This violates what the function documentation claims to be a valid return.

cachedout commented on salt/modules/mac_shadow.py:
```diff
@@ -189,8 +160,49 @@ def get_last_change(name):
+        pwd.getpwnam(name)
+    except KeyError:
+        log.debug('User not found: {0}'.format(name))
+        raise CommandExecutionError('User not found: {0}'.format(name))
```
If we're going to do this, we need to document it so that a user knows to expect it as an error condition.

cachedout commented:
This still need some work. Please see in-line comments.

twangboy committed:
Added documentation about raise error

Refactored

twangboy committed:
Syntax Error and Lint

twangboy committed:
Fix User not found detection

twangboy committed:
Fix get_login_failed_count function

twangboy committed:
Fix _set_account_policy function

twangboy committed:
Add more descriptive documentation for errors

cachedout commented on salt/modules/mac_shadow.py:
```diff
@@ -30,95 +33,99 @@ def __virtual__():
-    Returns:
-        the value found in the command
+    :return: a dictionary containing all values for the accountPolicy
+    :rtype: bool
```
Is it a dictionary or a bool?

cachedout commented on salt/modules/mac_shadow.py:
```diff
@@ -30,95 +33,99 @@ def __virtual__():
+    except CommandExecutionError as exc:
+        if 'Error: user <{0}> not found'.format(name) in exc.strerror:
+            raise CommandExecutionError('User not found: {0}'.format(name))
+        raise CommandExecutionError('Unknown error: {0}'.format(exc.strerror))
```
I'm not sure what good this block does. All it really does is rewrite the exception string.

twangboy replied:
It will either return a CommandExecutionError with 'User not found: <user_name>' as the message or 'Unknown error: <error_message>' on any other error.

cachedout replied:
I understand that. I'm just saying that the prefix doesn't seem to helpful since it's rewriting `Error: user <foo> not found` to `User not found: foo` and otherwise just adding a prefix to an exception string that would already be raised. But, it's fine. I'll merge it. :]

cachedout commented on salt/modules/mac_shadow.py:
```diff
@@ -260,66 +324,71 @@ def get_maxdays(name):
+    return False
 
 
 def set_warndays(name, days):
```
CLI example missing.

twangboy replied:
@cachedout Added documentation

twangboy committed:
Fix documentation

Fixed errorneous return type for _get_account_policy
Added documentation for unsupported functions
